core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -266,130 +267,148 @@ private static boolean isConflictException(final Exception e) {
+                        XContentHelper.convertToMap(indexSourceAsBytes, true, updateIndexRequest.getContentType());
+                updateResponse.setGetResult(UpdateHelper.extractGetResult(updateRequest, request.index(),
+                                indexResponse.getVersion(), sourceAndContent.v2(), sourceAndContent.v1(), indexSourceAsBytes));
+            }
+            // set translated request as replica request
+            BulkItemRequest replicaRequest = new BulkItemRequest(bulkReqId, updateIndexRequest);
+            return new BulkItemResultHolder(updateResponse, result, replicaRequest);
+
+        } else if (opType == Engine.Operation.TYPE.DELETE) {
+            assert result instanceof Engine.DeleteResult : result.getClass();
+            final DeleteRequest updateDeleteRequest = translate.action();
+
+            final DeleteResponse deleteResponse = new DeleteResponse(primary.shardId(), updateDeleteRequest.type(), updateDeleteRequest.id(),
+                    result.getSeqNo(), primary.getPrimaryTerm(), result.getVersion(), ((Engine.DeleteResult) result).isFound());
+
+            final UpdateResponse updateResponse = new UpdateResponse(deleteResponse.getShardInfo(), deleteResponse.getShardId(),
+                    deleteResponse.getType(), deleteResponse.getId(), deleteResponse.getSeqNo(), deleteResponse.getPrimaryTerm(),
+                    deleteResponse.getVersion(), deleteResponse.getResult());
+
+            final GetResult getResult = UpdateHelper.extractGetResult(updateRequest, request.index(), deleteResponse.getVersion(),
+                    translate.updatedSourceAsMap(), translate.updateSourceContentType(), null);
+
+            updateResponse.setGetResult(getResult);
+            // set translated request as replica request
+            BulkItemRequest replicaRequest = new BulkItemRequest(bulkReqId, updateDeleteRequest);
+            return new BulkItemResultHolder(updateResponse, result, replicaRequest);
+        } else {
+            throw new IllegalArgumentException("unknown operation type: " + opType);
+        }
+    }
+
+    /**
+     * Executes update request once, delegating to a index or delete operation after translation.
+     * NOOP updates are indicated by returning a <code>null</code> operation in {@link BulkItemResultHolder}
+     */
+    static BulkItemResultHolder executeUpdateRequestOnce(UpdateRequest updateRequest, IndexShard primary,
+                                                         IndexMetaData metaData, BulkShardRequest request,
+                                                         int requestIndex, UpdateHelper updateHelper,
+                                                         LongSupplier nowInMillis,
+                                                         final MappingUpdatePerformer mappingUpdater) throws Exception {
+        BulkItemRequest originalReplicaRequest = request.items()[requestIndex];
```

Comment:
can we call this primaryItemRequest? It's the one that's sent to the primary. Also, if we pass it as a `BulkItemRequest` parameter, we can avoid sending `requestIndex` and `BulkShardRequest` (from which need the concreteIndex, which I think we can from the shard). Last can we assert that the `BulkItemRequest` has as a request object the `updateRequest` we got? this is all super trappy but we can take one step at a time :)